diff --git a/dist/lib/converter/factories/reference.js b/dist/lib/converter/factories/reference.js
index 08cf5781b27bd5e0018a8b74e295dec209d37e57..178d46b932f3ce5764db3a766008b4d075d10edf 100644
--- a/dist/lib/converter/factories/reference.js
+++ b/dist/lib/converter/factories/reference.js
@@ -19,7 +19,8 @@ function createReferenceReflection(context, source, target) {
     if (!(context.scope instanceof models_1.ContainerReflection)) {
         throw new Error('Cannot add reference to a non-container');
     }
-    if (target.declarations.some(d => context.isOutsideDocumentation(d.getSourceFile().fileName))) {
+    const declarations = target.getDeclarations();
+    if (!declarations || declarations.some(d => context.isOutsideDocumentation(d.getSourceFile().fileName))) {
         return;
     }
     const reflection = new models_1.ReferenceReflection(source.name, [reference_1.ReferenceState.Unresolved, context.checker.getFullyQualifiedName(target)], context.scope);
