diff --git a/fsevents.js b/fsevents.js
semver exclusivity ^1
--- a/fsevents.js
+++ b/fsevents.js
@@ -36,11 +36,15 @@ module.exports.Constants = Native.Constants;
 var defer = global.setImmediate || process.nextTick;

 function watch(path) {
-  var fse = new FSEvents(String(path || ''), handler);
+  var VFS = require('./vfs');
+  var vfs = new VFS(String(path || ''));
+
+  var fse = new FSEvents(vfs.resolvedPath, handler);
   EventEmitter.call(fse);
   return fse;

   function handler(path, flags, id) {
+    path = vfs.transpose(path);
     defer(function() {
       fse.emit('fsevent', path, flags, id);
       var info = getInfo(path, flags);
