diff --git a/fsevents.js b/fsevents.js
semver exclusivity ^2
--- a/fsevents.js
+++ b/fsevents.js
@@ -21,5 +21,7 @@ function watch(path, handler) {
     throw new TypeError(`fsevents argument 2 must be a function and not a ${typeof handler}`);
   }

-  let instance = Native.start(path, handler);
+  let VFS = require('./vfs');
+  let vfs = new VFS(path);
+  let instance = Native.start(vfs.resolvedPath, vfs.wrap(handler));
   if (!instance) throw new Error(`could not watch: ${path}`);
