diff --git a/fsevents.js b/fsevents.js
semver exclusivity 2.1.1
--- a/fsevents.js
+++ b/fsevents.js
@@ -21,12 +21,11 @@ function watch(path, handler) {
     throw new TypeError(`fsevents argument 2 must be a function and not a ${typeof handler}`);
   }
 
-  let instance = Native.start(path, handler);
-  if (!instance) throw new Error(`could not watch: ${path}`);
+  let VFS = require('./vfs');
+  let vfs = new VFS(path, Native);
+  vfs.watch(handler);
   return () => {
-    const result = instance ? Promise.resolve(instance).then(Native.stop) : null;
-    instance = null;
-    return result;
+    return vfs.stop();
   };
 }
 
