diff --git a/vfs.js b/vfs.js
new file mode 100644
--- /dev/null
+++ b/vfs.js
@@ -0,0 +1,41 @@
+const path = require(`path`);
+
+const NUMBER_REGEXP = /^[0-9]+$/;
+const VIRTUAL_REGEXP = /^(\/(?:[^\/]+\/)*?\$\$virtual)((?:\/([^\/]+)(?:\/([^\/]+))?)?((?:\/.*)?))$/;
+
+function resolveVirtual(p) {
+  const match = p.match(VIRTUAL_REGEXP);
+  if (!match)
+    return p;
+
+  const target = path.dirname(match[1]);
+  if (!match[3] || !match[4])
+    return target;
+
+  const isnum = NUMBER_REGEXP.test(match[4]);
+  if (!isnum)
+    return p;
+
+  const depth = Number(match[4]);
+  const backstep = `../`.repeat(depth);
+  const subpath = (match[5] || `.`);
+
+  return resolveVirtual(path.join(target, backstep, subpath));
+}
+
+module.exports = class FsePnp {
+  constructor(p) {
+    this.normalizedPath = path.resolve(p);
+    this.resolvedPath = resolveVirtual(this.normalizedPath);
+  }
+
+  transpose(p) {
+    return this.normalizedPath + p.substr(this.resolvedPath.length);
+  }
+
+  wrap(fn) {
+    return (path, ...args) => {
+      return fn(this.transpose(path), ...args);
+    };
+  }
+};
